### PR TITLE
Tokenizer: prefer HF Tokenizer

### PR DESCRIPTION
**Pull request by Andrei-Aksionov**
Hi there 👋 

This PR changes the default tokenizer from SentencePiece to Tokenizers by HF.
There are mainly three reasons why it's a good idea:
1. Tokenizers library is faster* than SentencePiece
2. Easier to maintain
3. Easier to extend (if necessary)

> [!IMPORTANT]
> If a PR is created from a fork, then the `HF_TOKEN` is not shared. That means that tests for models/tokenizers from gated repos are skipped. **Before merging someone from the core team has to copy this branch to a new one, create a PR and run tests.** Like it was done in #1342.

### 1. Speed

Let's validate it with a simple benchmark:

```python
from time import perf_counter

import requests
from huggingface_hub import snapshot_download
from sentencepiece import SentencePieceProcessor
from tokenizers import Tokenizer as HFTokenizer
from transformers import PreTrainedTokenizerFast

# download tokenizer vocabs
snapshot_download(repo_id="TinyLlama/TinyLlama-1.1B-Chat-v1.0", local_dir=".", allow_patterns=["tokenizer.json", "tokenizer.model"])

# download data to encode
data = requests.get("https://raw.githubusercontent.com/karpathy/ng-video-lecture/master/input.txt").text
print(f"Number of chars to encode: {len(data):,}")
print("-" * 50)

# tokenizers to benchmark
sentencepiece_tokenizer = SentencePieceProcessor(model_file="tokenizer.model")
hf_tokenizer = HFTokenizer.from_file(path="tokenizer.json")
fast_tokenizer = PreTrainedTokenizerFast(tokenizer_file="tokenizer.json")

def bench(func: callable, repeats: int = 30) -> float:
    start = perf_counter()
    for _ in range(repeats):
        func()
    return (perf_counter() - start) / repeats


print("SentencePiece encode time: ", bench(lambda: sentencepiece_tokenizer.encode(data)))
print("HF Tokenizer  encode time: ", bench(lambda: hf_tokenizer.encode(data)))
print("FastTokenizer encode time: ", bench(lambda: fast_tokenizer.encode(data)))
print("-" * 50)

sp_ids = sentencepiece_tokenizer.encode(data)
hf_ids = hf_tokenizer.encode(data).ids
ft_ids = fast_tokenizer.encode(data)

print("SentencePiece decode time: ", bench(lambda: sentencepiece_tokenizer.decode(sp_ids)))
print("HF Tokenizer  decode time: ", bench(lambda: hf_tokenizer.decode(hf_ids)))
print("FastTokenizer decode time: ", bench(lambda: fast_tokenizer.decode(hf_ids)))
```

```text
Number of chars to encode: 1,115,394
--------------------------------------------------
SentencePiece encode time:  1.2004516734004331
HF Tokenizer  encode time:  0.5400692415996067
FastTokenizer encode time:  0.5723549843329238
--------------------------------------------------
SentencePiece decode time:  0.30017661056675327
HF Tokenizer  decode time:  0.39618176593309423
FastTokenizer decode time:  2.1387836453999625
```

As one can see, `HF Tokenizer` is ~2.3 times faster to encode, but ~1.32 times slower to decode (kinda odd).
Anyway, from my POV, given that the focus of LitGPT is training, it's a worthy change.

### 2. Maintenance

Vocabs for Tokenizers (`tokenizer.json`) usually has a full set of tokens in comparison to SentencePiece (`tokenizer.model`)

### 3. Extensibility

If the vocab is not full (missing added tokens) and we need to extend it, then it's way-way easier to do with Tokenizers (it has `.add_tokens` method). With SentencePiece we need to first read the file (which is in a protobuf format btw) with a special tool, add additional tokens with corresponding weights, save the file, and only then open it in SentencePiece processor. Should be doable, but not well documented.

For those who are curious, AutoTokenizer from HF doesn't use SentencePiece for encoding/decoding. If only an SP vocab is provided (or `is_slow=True` argument is specified), then the code first loads the vocab into the SP tokenizer, then the vocab is extracted and passed to the HF Tokenizer instance.

### CODE REVIEW

**Comment by rasbt** [comment]
Thanks for the PR! Regarding the slower decoding, just a thought, but maybe the fast tokenizer variants (https://huggingface.co/docs/transformers/main/en/fast_tokenizers) could help. I am not sure though how many tokenizer variants this approach covers/supports yet.

**Comment by Andrei-Aksionov** [comment]
@rasbt I updated the benchmark code.

In HF notation:
- SentencePiece is a slow tokenizer
- HF Tokenizers  is a fast one

`PretrainedTokenizerFast` just uses `Tokenizers` underneath and adds additional logic.
Thus the `encode` speed is the same, but `decode` is much slower (it's the result of additional logic).

File: litgpt/tokenizer.py [hunk 1]
```diff
@@ -97,7 +101,8 @@ def encode(
             bos_id = self.bos_id
             if bos_id is None:
                 raise NotImplementedError("This tokenizer does not have a defined a bos token")
-            tokens = [bos_id] + tokens
+            if tokens[0] != bos_id:
+                tokens = [bos_id] + tokens
         if eos:
             tokens = tokens + [self.eos_id]
```
Comment (rasbt) [inline litgpt/tokenizer.py:107]: Should we add a check similar to the one above?

```suggestion
        if tokens and tokens[-1] != eos_id:
            tokens = tokens + [eos_id]
```

**Review by rasbt — APPROVED** [review]
Thanks for the PR. Looks great overall!

File: litgpt/tokenizer.py (hunk 1 above)
Comment (Andrei-Aksionov) [reply litgpt/tokenizer.py:107]: The issue was only with `bos` token.
But I guess it won't hurt.

Could you do the honors?
I'm already away from my laptop.

Comment (Andrei-Aksionov) [reply litgpt/tokenizer.py:107]: Just change your code a bit.
You want to add eos token only if it's set to True and it hasn't been added already.

Comment (rasbt) [reply litgpt/tokenizer.py:107]: Sounds good. What about the following update code (above) then?

Comment (Andrei-Aksionov) [reply litgpt/tokenizer.py:107]: `not tokens` is True for an empty list but also for a None value, so that means that the code might try to sum None and a list (if tokens is None), which will result in an error.

**Comment by Andrei-Aksionov** [comment]
Just remember that before merging someone needs to create a new branch in this repo (that is a copy of this branch), so CI will use HF_TOKEN and thus no tokenizer tests are skipped.


Or we can merge and hope for the best 😜

File: litgpt/tokenizer.py (hunk 1 above)
Comment (rasbt) [reply litgpt/tokenizer.py:107]: Ah my bad, I thought it would be an empty list. But then we would have that issue in the original code,

```
if eos:
    tokens = tokens + [self.eos_id]
```

too?

Comment (Andrei-Aksionov) [reply litgpt/tokenizer.py:107]: True. Keep your code, but add also a check that the processor didn't return a None value. If it's so, then it's definitely an error.

Comment (rasbt) [reply litgpt/tokenizer.py:107]: You mean adding an

```
if tokens is None:
    raise ValueError("`tokens` is None")
```

?

Comment (Andrei-Aksionov) [reply litgpt/tokenizer.py:107]: Yes 👍

**Comment by rasbt** [comment]
> Just remember that before merging someone needs to create a new branch in this repo (that is a copy of this branch), so CI will use HF_TOKEN and thus no tokenizer tests are skipped.

Oh I got it 😊 https://github.com/Lightning-AI/litgpt/pull/1522

File: litgpt/tokenizer.py [hunk 2]
```diff
@@ -97,8 +101,12 @@ def encode(
...
             if bos_id is None:
                 raise NotImplementedError("This tokenizer does not have a defined a bos token")
-            tokens = [bos_id] + tokens
-        if eos:
+            if tokens[0] != bos_id:
+                tokens = [bos_id] + tokens
+        if tokens is None:
+            raise ValueError("`tokens` is None")
+
+        if eos and (not tokens or tokens[-1] != eos_id):
```
Comment (rasbt) [inline litgpt/tokenizer.py:109]:
```suggestion
        if eos and (not tokens or tokens[-1] != self.eos_id):
```

File: litgpt/tokenizer.py [hunk 3]
```diff
@@ -97,8 +101,12 @@ def encode(
...
             if bos_id is None:
                 raise NotImplementedError("This tokenizer does not have a defined a bos token")
-            tokens = [bos_id] + tokens
-        if eos:
+            if tokens[0] != bos_id:
+                tokens = [bos_id] + tokens
+        if tokens is None:
+            raise ValueError("`tokens` is None")
+
+        if eos and (not tokens or tokens[-1] != self.eos_id):
```
Comment (awaelchli) [inline litgpt/tokenizer.py:109]: Could you add or extend a test to cover this new behavior? Might also be cleaner to do this in a separate PR.

Comment (Andrei-Aksionov) [reply litgpt/tokenizer.py:109]: Yes, good idea.
I'll make a separate PR.

Comment (rasbt) [reply litgpt/tokenizer.py:109]: Sounds good @Andrei-Aksionov . I can merge then and you could do that in a separate PR then like you said.
